### PR TITLE
add progress bar indeterminate mode

### PR DESCRIPTION
**Pull request by ektasawant**
Fixes # .
PXBLUE-2148

#### Changes proposed in this Pull Request:
-  Spinner overlay
- Contextual Spinner
- Progress bar
- Progress bar (inte)
- Skeleton Loader

### CODE REVIEW

File: src/app/pages/loading-waiting-states/progress-bar-indeterminate/progress-bar-indeterminate.component.html [hunk 1]
```diff
@@ -0,0 +1,14 @@
+<mat-toolbar data-cy="pxb-toolbar" flexLayout="row" class="app-bar mat-elevation-z2" [color]="checked ? 'primary' : ''">
```
Comment (huayunh) [inline src/app/pages/loading-waiting-states/progress-bar-indeterminate/progress-bar-indeterminate.component.html:1]: mat-elevation-z4 is the usual elevation we use for app bar

File: src/app/pages/loading-waiting-states/progress-bar-indeterminate/progress-bar-indeterminate.component.html [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+    <button data-cy="toolbar-menu" *ngIf="isSmall" mat-icon-button (click)="toggleMenu()" style="margin-left: -8px">
+        <mat-icon>menu</mat-icon>
+    </button>
+    <div class="mat-title" [style.marginLeft.px]="isSmall ? 24 : 0">Progress Bars (Indeterminate)</div>
+    <mat-progress-bar mode="indeterminate"></mat-progress-bar>
+    <pxb-spacer></pxb-spacer>
+    <mat-slide-toggle labelPosition="before" [checked]="checked" (change)="onToggle($event)">Use Dark Theme</mat-slide-toggle>
```
Comment (huayunh) [inline src/app/pages/loading-waiting-states/progress-bar-indeterminate/progress-bar-indeterminate.component.html:8]: Use dark theme should be a global functionality — i.e., toggling the button in one page would change the overall application state

Comment (huayunh) [reply src/app/pages/loading-waiting-states/progress-bar-indeterminate/progress-bar-indeterminate.component.html:8]: On mobile, the toggle would become a menu item — see the leftmost screen in Figma

Comment (huayunh) [reply src/app/pages/loading-waiting-states/progress-bar-indeterminate/progress-bar-indeterminate.component.html:8]: I am happy to leave this functionality out for this round though. We can revisit this in the future.

File: src/app/pages/loading-waiting-states/progress-bar-indeterminate/progress-bar-indeterminate.component.html [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+</mat-toolbar>
+<section class="mat-typography" [class.small]="isSmall">
```
Comment (huayunh) [inline src/app/pages/loading-waiting-states/progress-bar-indeterminate/progress-bar-indeterminate.component.html:10]: Page body shall be white200

**Review by huayunh — COMMENTED** [review]
Visually looks pretty good! Remember to run `yarn prettier` before you push

File: src/app/pages/loading-waiting-states/progress-bar-indeterminate/progress-bar-indeterminate.component.html (hunk 1 above)
Comment (ektasawant) [reply src/app/pages/loading-waiting-states/progress-bar-indeterminate/progress-bar-indeterminate.component.html:1]: Sure let me update.

File: src/app/pages/loading-waiting-states/progress-bar-indeterminate/progress-bar-indeterminate.component.html (hunk 3 above)
Comment (joebochill) [reply src/app/pages/loading-waiting-states/progress-bar-indeterminate/progress-bar-indeterminate.component.html:10]: this should be a global style somewhere for the app shouldn't it? Shouldn't this come from our theme?

File: src/app/pages/loading-waiting-states/contextual-spinners/contextual-spinners.component.ts [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+@Component({
+    selector: 'app-contextual-spinners',
+    templateUrl: './contextual-spinners.component.html',
+    styleUrls: ['./contextual-spinners.component.scss'],
+})
+export class ContextualSpinnersComponent implements OnInit {
+    isSmall: boolean;
+    checked = false;
+    buttonClicked = false;
+    flatbuttonClicked = false;
```
Comment (daileytj) [inline src/app/pages/loading-waiting-states/contextual-spinners/contextual-spinners.component.ts:15]: This is really minor, but can you update this variable to `flatButtonClicked` throughout this example to maintain the camelcase format.

File: src/app/pages/loading-waiting-states/contextual-spinners/contextual-spinners.component.html [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    <div class="button-examples-container">
+        <h5>Click on the buttons below to see the spinners.</h5>
+        <div class="example-section extended-fab-button-example">
+            <button class="extended-fab-button" mat-fab color="primary" (click)="loadSpinner()" *ngIf="!buttonClicked">
+                <mat-icon>play_arrow</mat-icon>
+                <span class="extended-fab-button__text">Start Routine</span>
+            </button>
+            <button *ngIf="buttonClicked" class="button-container" mat-fab color="primary">
+                <mat-spinner class="primary-spineer" diameter="20"></mat-spinner>
+            </button>
```
Comment (daileytj) [inline src/app/pages/loading-waiting-states/contextual-spinners/contextual-spinners.component.html:23]: @huayunh Should this button be disabled when the spinner is showing?

File: src/app/pages/loading-waiting-states/contextual-spinners/contextual-spinners.component.html [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+                Log In
+            </button>
+            <button
+                mat-raised-button
+                class="button-container raised-button-example"
+                color="primary"
+                *ngIf="flatbuttonClicked"
+            >
+                <mat-spinner class="primary-spinner" diameter="20"></mat-spinner>
+            </button>
```
Comment (daileytj) [inline src/app/pages/loading-waiting-states/contextual-spinners/contextual-spinners.component.html:36]: @huayunh Same question here...

File: src/app/pages/loading-waiting-states/progress-bar-indeterminate/progress-bar-indeterminate.module.ts [hunk 7]
```diff
@@ -0,0 +1,25 @@
+import { NgModule } from '@angular/core';
+import { CommonModule } from '@angular/common';
+import { ProgressBarIndeterminateComponent } from './progress-bar-indeterminate.component';
+import { MatToolbarModule } from '@angular/material/toolbar';
+import { MatButtonModule } from '@angular/material/button';
+import { MatIconModule } from '@angular/material/icon';
+import { MatProgressBarModule } from '@angular/material/progress-bar';
+import { MatSlideToggleModule } from '@angular/material/slide-toggle';
+import { SpacerModule } from '@pxblue/angular-components';
+import { EmptyStateModule } from '@pxblue/angular-components';
```
Comment (daileytj) [inline src/app/pages/loading-waiting-states/progress-bar-indeterminate/progress-bar-indeterminate.module.ts:10]: You have duplicate import statements for angular-components. 

```suggestion
import { SpacerModule, EmptyStateModule } from '@pxblue/angular-components';
```

File: src/app/pages/loading-waiting-states/progress-bars/progress-bars.module.ts [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+import { MatToolbarModule } from '@angular/material/toolbar';
+import { MatCardModule } from '@angular/material/card';
+import { MatButtonModule } from '@angular/material/button';
+import { MatIconModule } from '@angular/material/icon';
+import { MatProgressBarModule } from '@angular/material/progress-bar';
+import { MatSlideToggleModule } from '@angular/material/slide-toggle';
+import { MatRadioModule } from '@angular/material/radio';
+import { MatDividerModule } from '@angular/material/divider';
+import { SpacerModule } from '@pxblue/angular-components';
+import { InfoListItemModule } from '@pxblue/angular-components';
```
Comment (daileytj) [inline src/app/pages/loading-waiting-states/progress-bars/progress-bars.module.ts:13]: Duplicate import from angular-components...

```suggestion
import { SpacerModule, InfoListItemModule } from '@pxblue/angular-components';
```

File: src/app/pages/loading-waiting-states/progress-bars/progress-bars.component.ts [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+  toggleMenu(): void {
+    const drawerOpen = this._drawerService.getDrawerOpen();
+    this._drawerService.setDrawerOpen(!drawerOpen);
+  }
+
+  onToggle(event: MatSlideToggleChange): void {
+    this.checked = event.checked;
+  }
+
+  uploadFile() {
```
Comment (daileytj) [inline src/app/pages/loading-waiting-states/progress-bars/progress-bars.component.ts:51]: Missing return type

```suggestion
  uploadFile(): void {
```

File: src/app/pages/loading-waiting-states/progress-bars/progress-bars.component.ts [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+    this.uploadFileList.push({ fileName: 'PX Blue is Awesome.pdf', progress: 0 })
+    for (let i = 0; i <= this.uploadFileList.length; i++) {
+      this.interval = setInterval(() => {
```
Comment (daileytj) [inline src/app/pages/loading-waiting-states/progress-bars/progress-bars.component.ts:54]: Missing return type here too.

File: src/app/pages/loading-waiting-states/progress-bar-indeterminate/progress-bar-indeterminate.component.html (hunk 3 above)
Comment (ektasawant) [reply src/app/pages/loading-waiting-states/progress-bar-indeterminate/progress-bar-indeterminate.component.html:13]: I couldn't find global style for it.

**Comment by joebochill** [comment]
Couple of comments:

1. Spinner Overlays
    - The spinner should be in an overlay with a semi-transparent background color (skrim) that covers all of the contents of the card, rather than just block level content within the card.
2. Contextual Spinners
    - Question for @huayunh: do  we want the buttons to shrink when they have a spinner, or should they preserve their original size?
3. Skeletons
    - Looking good so far...looks like you're still working on this one.
4. Progress Bars
    - Looks really nice
    - There are errors getting logged in the console: 
    image
    - We should be able to remove the snackbars by clicking the cancel button
    - @huayunh do we want to auto-remove items after a period of time once the upload is complete or just manually?

**Comment by huayunh** [comment]
## Spinner Overlays

- The divider above the reload button should span full width across the container
- Increase the height of "End User License Agreement"'s `div` 

## Contextual Spinners
- When a button is transitioning its state into a spinner state, do use some transition animation (shrink and fade out existing content, expand and fade in the new content, button width should also be animated)
- For the log in button, preserve its width
- note the margin-left of expanded app bar
image

## Progress Bars
- I have just updated the instruction on Figma — basically, let the snackbar dismiss itself 3s after a file is uploaded
- Use `text_snippet` for the icon in the snackbar
- Distance between radio button, the folder icon, and their label should be 16px
- The folder icon is sized at 18px and gray500

## Progress Bars (Indeterminate)
@emclaug2 can you give some instructions on how to get the default background color (white200)?

## Other
- Run `yarn prettier` and `yarn lint:fix` before you push
- Remember to remove the "loading states" (page highlighted below) once you are finished with everything
  image

**Comment by joebochill** [comment]
Some updated feedback:
- Spinner overlay looks good
- Contextual Spinners
    - collapsing fab collapse/expand animation feels a tad bit slow
    - don't change the border radius in the animation (when you use 50%, it turns into an oval in the middle instead of a rounded rectangle)
    - Skeletons: use skeletons over the "View Location" part of the scorecard and the right arrow in the InfoListItems
    - Progress Bars looks great
    - Indeterminate progress bars looks great

**Comment by emclaug2** [comment]
> @emclaug2 can you give some instructions on how to get the default background color (white200)?

in styles.scss, line 16 there is a style that is providing a bg color to the the page content.
```
.pxb-drawer-layout-nav-content {
    background-color: map-get($pxb-white, 50);
}
```

File: src/app/pages/loading-waiting-states/progress-bar-indeterminate/progress-bar-indeterminate.component.html (hunk 3 above)
Comment (joebochill) [reply src/app/pages/loading-waiting-states/progress-bar-indeterminate/progress-bar-indeterminate.component.html:13]: @emclaug2 will know

Comment (joebochill) [reply src/app/pages/loading-waiting-states/progress-bar-indeterminate/progress-bar-indeterminate.component.html:13]: nvm, I see he already commented below

File: src/app/pages/loading-waiting-states/progress-bars/progress-bars.component.scss [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+        text-align: right;
+        padding-bottom: 16px;
+    }
+    .folder-lists-container {
+        padding: 0;
+    }
+    .radio-group {
+        display: flex;
+        flex-direction: column;
+        margin: 15px 0;
```
Comment (joebochill) [inline src/app/pages/loading-waiting-states/progress-bars/progress-bars.component.scss:17]: We generally try to do all of our spacings (margins, paddings, sizes) as multiples of 8px.
```suggestion
        margin: 16px 0;
```

File: src/app/pages/loading-waiting-states/progress-bars/progress-bars.component.scss [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+    }
+
+    .folder-name {
+        display: flex;
+        align-items: center;
+
+        .mat-icon {
+            margin: 0 16px 6px 4px;
```
Comment (joebochill) [inline src/app/pages/loading-waiting-states/progress-bars/progress-bars.component.scss:25]: the 6 here is a little odd - can we make it work on the 8px grid?

File: src/app/pages/loading-waiting-states/contextual-spinners/contextual-spinners.component.scss [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+.extended-fab-button {
+    width: 180px;
+    height: 56px;
+    padding: 0 26px;
+    border-radius: 29px;
+    transition-property: unset;
+    transition-duration: 1s;
+
+    .mat-icon {
+        margin-right: 10px;
```
Comment (joebochill) [inline src/app/pages/loading-waiting-states/contextual-spinners/contextual-spinners.component.scss:12]: Can we make this 8px to align on the 8px grid?

File: src/app/pages/loading-waiting-states/contextual-spinners/contextual-spinners.component.scss [hunk 14]
```diff
@@ -0,0 +1,53 @@
+@import '~@pxblue/colors/palette.scss';
+
+.extended-fab-button {
+    width: 180px;
+    height: 56px;
+    padding: 0 26px;
```
Comment (joebochill) [inline src/app/pages/loading-waiting-states/contextual-spinners/contextual-spinners.component.scss:6]: The padding here is a little odd, but I can see why you did it that way. I think it's fine for now.

File: src/app/pages/loading-waiting-states/progress-bars/progress-bars.component.ts [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+export class ProgressBarsComponent implements OnInit {
+    isSmall: boolean;
+    checked = false;
+    foldersList = [
+        { label: 'The Best Dev Team', value: '1' },
+        { label: 'The Best Design Team', value: '2' },
+        { label: 'The Best UX Team', value: '3' },
+        { label: 'The Best Management Team', value: '4' },
+        { label: 'The Best Facility Team', value: '5' },
+        { label: 'The Most Narcissist Team', value: '6' },
```
Comment (joebochill) [inline src/app/pages/loading-waiting-states/progress-bars/progress-bars.component.ts:19]:
```suggestion
        { label: 'The Proudest Team', value: '6' },
```

File: src/app/pages/loading-waiting-states/skeleton-loader/components/placeholder-hero/placeholder-hero.component.scss [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+        height: 140px;
+        .ph-item {
+            padding-bottom: 0;
+            .ph-col-4 {
+                justify-content: center;
+                max-width: 100px;
+                .ph-avatar {
+                    min-width: 0;
+                    height: 36px;
+                    width: 36px;
```
Comment (joebochill) [inline src/app/pages/loading-waiting-states/skeleton-loader/components/placeholder-hero/placeholder-hero.component.scss:16]: The avatars are usually 40px I believe

File: src/app/pages/loading-waiting-states/skeleton-loader/components/placeholder-score-card/placeholder-score-card.component.scss [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+                height: 24px;
+                width: 24px;
+                margin-bottom: 8px;
+            }
+        }
+
+        .notification-section {
+            .ph-col-12 {
+                height: 14px;
+                margin: 5px 0px;
```
Comment (joebochill) [inline src/app/pages/loading-waiting-states/skeleton-loader/components/placeholder-score-card/placeholder-score-card.component.scss:53]:
```suggestion
                margin: 4px 0px;
```

**Comment by ektasawant** [comment]
Shall I update following

`.pxb-drawer-layout-nav-content {
    background-color: map-get($pxb-white, 50);
}`  => `.pxb-drawer-layout-nav-content {
    background-color: map-get($pxb-white, 200);
}`

File: src/app/pages/loading-waiting-states/spinner-overlays/spinner-overlays.component.scss [hunk 18]
```diff
@@ -0,0 +1,92 @@
...
+        }
+    }
+
+    .eula-content {
+        padding-top: 8px;
+    }
+
+    .eula-divider.mat-divider-horizontal {
+        position: relative;
+        margin: 15px 0 8px;
```
Comment (joebochill) [inline src/app/pages/loading-waiting-states/spinner-overlays/spinner-overlays.component.scss:81]:
```suggestion
        margin: 16px 0 8px;
```

**Review by joebochill — COMMENTED** [review]
Scorecard hero icons aren't the right size:
image

For the skeleton list items can we get a placeholder for the chevron on the right (or hide it when we are loading)?
image

**Review by joebochill — APPROVED** [review]

**Pull request merged**